docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,200 @@
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases when a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+* Do not run an agent on a node if it is not being unbound
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+* Registering an agent to an InfraEnv other than the one it is currently
+  associated with
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host as a systemd service
+using a MachineConfig. This agent will handle downloading the required
+artifacts and rebooting the host into the discovery process.
+
+This process will only be used in situations where the assisted-service is sure
+to have connectivity to the installed cluster's API endpoint. In cases such as
+the SaaS this feature will not be used.
+
+### Timeline
+
+1. User unbinds the host from the cluster
+2. Assisted service creates a MachineConfig to run the assisted-installer-agent
+   on the host as a systemd service
+  * If there is no BareMetalHost associated with the agent CR, the service
+    executes a new host state transition (`TransitionTypeReclaimHost`)
+  * Host moves to state `HostStatusReclaiming`
+3. The agent downloads the InfraEnv's kernel and initrd (including discovery
+   ignition) to the host's `/boot` partition
+4. The agent creates a GRUB boot entry to boot into the discovery kernel and
+   initrd
+5. The agent reboots the host into the discovery image
+6. Discovery agent registers host
+  * Service moves host to `HostStatusDiscoveringUnbound`
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts
+from existing clusters without requiring access to the host management
+infrastructure (BMC, hypervisor, etc).
+
+#### Story 2
+
+As a cluster creator, I want the host's storage wiped when it is removed from a
```

Comment:
Data also for cases where the users of a cluster-as-a-service want to prevent data leakage to other users. Or even just ensure other drives are in a consistent clean state for discovery and consumption by thing like ODF.

Reply:
I don't know if I would include this story in the enhancement.  It's a valid story regardless of how the host was re-used.

Reply:
The important part is that for some use cases, this will be a hard requirement, and it could be a serious breech if data is leaked. "automating returning an agent to the infra-env" for those use cases would have to include wiping data. If we ship this feature without that, it needs to be loud and clear that data is not wiped.

Reply:
Maybe (for the next iteration) there could be an exposed option to the user where they could choose whether or not to wipe the data on certain/all hosts?

Reply:
Maybe, but we don't really have a way for users to provide us with any additional information right now. Especially when unbinding in the kube API.

I think that something like this might require a bit more design.